### PR TITLE
CPM-1027: Get and Update Ref entities nomenclatures

### PR DESCRIPTION
**Pull request by cemorin**
**Description (for Contributor and Core Developer)**

Handle ref entities in GetNomenclatureController and UpdateNomenclatureController

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                'values' => [
+                    'akeneo' => 'akn',
+                    'adidas' => 'adds',
+                    'zara' => '',
+                    'unknown' => 'TOTO',
+                ],
+            ]),
+        );
+        $response = $this->client->getResponse();
+        Assert::assertSame(Response::HTTP_OK, $response->getStatusCode());
```
Comment (Beninho) [inline components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php:69]: You have to swap the value and what is expected

```suggestion
        Assert::assertSame($response->getStatusCode(), Response::HTTP_OK);
```

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                'value' => 4,
+                'generate_if_empty' => true,
+                'values' => [
+                    'adidas' => null,
+                    'zara' => 'zr',
+                ],
+            ]),
+        );
+        $response = $this->client->getResponse();
+        Assert::assertSame(Response::HTTP_OK, $response->getStatusCode());
```
Comment (Beninho) [inline components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php:95]:
```suggestion
        Assert::assertSame($response->getStatusCode(), Response::HTTP_OK);
```

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            ['HTTP_X-Requested-With' => 'XMLHttpRequest'],
+            \json_encode([
+                'operator' => '<=',
+                'value' => 4,
+                'generate_if_empty' => true,
+                'values' => [],
+            ]),
+        );
+        $response = $this->client->getResponse();
+        Assert::assertSame(Response::HTTP_NOT_FOUND, $response->getStatusCode());
```
Comment (Beninho) [inline components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php:119]:
```suggestion
        Assert::assertSame($response->getStatusCode(), Response::HTTP_NOT_FOUND);
```

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    public function it_should_redirect_on_non_xhr_request(): void
+    {
+        $this->loginAs('Julia');
+        $this->callUpdateRoute(
+            'akeneo_identifier_generator_nomenclature_rest_update',
+            ['propertyCode' => 'a_reference_entity_attribute'],
+            ['HTTP_X-Requested-With' => 'toto']
+        );
+        $response = $this->client->getResponse();
+        Assert::assertSame(Response::HTTP_FOUND, $response->getStatusCode());
```
Comment (Beninho) [inline components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php:43]:
```suggestion
        Assert::assertSame($response->getStatusCode(), Response::HTTP_FOUND);
```

**Review by Beninho — APPROVED** [review]

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php (hunk 1 above)
Comment (mmetayer) [reply components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php:69]: nope, in phpunit's assertions, expected value is the first argument

File: components/identifier-generator/back/src/Application/Exception/UndefinedAttributeException.php [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+declare(strict_types=1);
+
+namespace Akeneo\Pim\Automation\IdentifierGenerator\Application\Exception;
+
+/**
+ * @copyright 2023 Akeneo SAS (https://www.akeneo.com)
+ * @license   https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class UndefinedAttributeException extends \Exception
```
Comment (mmetayer) [inline components/identifier-generator/back/src/Application/Exception/UndefinedAttributeException.php:11]:
```suggestion
class UndefinedAttributeException extends \RuntimeException
```

File: components/identifier-generator/back/src/Application/Get/GetNomenclatureHandler.php [hunk 6]
```diff
@@ -35,7 +41,18 @@ public function __invoke(GetNomenclatureCommand $command): array
...
+            if (null === $attribute) {
+                throw new UndefinedAttributeException(
+                    \sprintf('The "%s" attribute is not found', $command->propertyCode())
+                );
+            }
+            if ($attribute->type() === AttributeTypes::OPTION_SIMPLE_SELECT) {
+                $nomenclature = $this->simpleSelectNomenclatureRepository->get($command->propertyCode()) ?? new NomenclatureDefinition();
+            } else {
+                $nomenclature = $this->referenceEntityNomenclatureRepository->get($command->propertyCode()) ?? new NomenclatureDefinition();
+            }
```
Comment (mmetayer) [inline components/identifier-generator/back/src/Application/Get/GetNomenclatureHandler.php:55]:
```suggestion
            $nomenclature = match($attribute->type()) {
                AttributeTypes::OPTION_SIMPLE_SELECT => $this->simpleSelectNomenclatureRepository->get($command->propertyCode()) ?? new NomenclatureDefinition(),
                AttributeTypes::REFERENCE_ENTITY_SIMPLE_SELECT => $this->referenceEntityNomenclatureRepository->get($command->propertyCode()) ?? new NomenclatureDefinition(),
                default => throw new UnexpectedAttributeTypeException();
            };
```

File: components/identifier-generator/back/src/Infrastructure/Controller/GetNomenclatureController.php [hunk 7]
```diff
@@ -30,6 +32,10 @@ public function __invoke(Request $request, string $propertyCode): Response
 
         $command = new GetNomenclatureCommand($propertyCode);
```
Comment (mmetayer) [inline components/identifier-generator/back/src/Infrastructure/Controller/GetNomenclatureController.php:33]: s/Command/Query/ (in another PR, can you create a card for it please?)

File: components/identifier-generator/back/src/Infrastructure/Controller/GetNomenclatureController.php [hunk 8]
```diff
@@ -30,6 +32,10 @@ public function __invoke(Request $request, string $propertyCode): Response
...
 
-        return new JsonResponse(($this->getNomenclatureHandler)($command));
+        try {
+            return new JsonResponse(($this->getNomenclatureHandler)($command));
+        } catch (UndefinedAttributeException $e) {
+            throw new NotFoundHttpException($e->getMessage());
```
Comment (mmetayer) [inline components/identifier-generator/back/src/Infrastructure/Controller/GetNomenclatureController.php:38]: I'm not really a fan of directly passing the exception message to the frontend. Also, what happens in th UI when receiving a 404?

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+                ],
+            ]),
+        );
+        $response = $this->client->getResponse();
+        Assert::assertSame(Response::HTTP_OK, $response->getStatusCode());
+
+        $nomenclatureDefinition = $this->getNomenclatureRepository()->get('a_reference_entity_attribute');
+        Assert::assertSame(($nomenclatureDefinition->values() ?? [])['akeneo'] ?? null, 'akn');
+        Assert::assertSame(($nomenclatureDefinition->values() ?? [])['adidas'] ?? null, null);
+        Assert::assertSame(($nomenclatureDefinition->values() ?? [])['zara'] ?? null, 'zr');
```
Comment (mmetayer) [inline components/identifier-generator/back/tests/EndToEnd/Infrastructure/Controller/UpdateReferenceEntityNomenclatureControllerEndToEnd.php:100]: arguments should be swapped here

File: components/identifier-generator/back/src/Infrastructure/Controller/GetNomenclatureController.php (hunk 7 above)
Comment (cemorin) [reply components/identifier-generator/back/src/Infrastructure/Controller/GetNomenclatureController.php:33]: Ticket created ! :+1:

File: components/identifier-generator/back/src/Infrastructure/Controller/GetNomenclatureController.php (hunk 8 above)
Comment (cemorin) [reply components/identifier-generator/back/src/Infrastructure/Controller/GetNomenclatureController.php:40]: Seen IRL, ticket created to handle it properly on the frontend side: https://akeneo.atlassian.net/browse/CPM-1059

File: components/identifier-generator/back/src/Application/Update/UpdateNomenclatureHandler.php [hunk 10]
```diff
@@ -21,18 +27,26 @@ public function __construct(
...
 
-        if ($command->getPropertyCode() === FamilyProperty::TYPE) {
-            $nomenclatureDefinition = $this->familyNomenclatureRepository->get() ?? new NomenclatureDefinition();
-        } else {
-            $nomenclatureDefinition = $this->simpleSelectNomenclatureRepository->get($command->getPropertyCode()) ?? new NomenclatureDefinition();
-        }
+        $getNomenclatureCommand = new GetNomenclatureCommand($command->getPropertyCode());
+        $normalizedNomenclature = ($this->getNomenclatureHandler)($getNomenclatureCommand);
+
+        // TODO: not sure we have to get existing nomenclature as we re-construct it wholy line 55
```
Comment (cemorin) [inline components/identifier-generator/back/src/Application/Update/UpdateNomenclatureHandler.php:43]: To check! @mmetayer what's your point of view about this? (we only send values we want to update, not all of them)

**Review by mmetayer — APPROVED** [review]

**Pull request merged**